tools/wptrunner/wptrunner/executors/executorchrome.py

```diff
@@ -57,12 +61,86 @@ def convert_from_crashtest_result(test, result):
+    that the test window can be reused between tests for better performance.
+    """
 
-class ChromeDriverTestharnessExecutor(WebDriverTestharnessExecutor, _SanitizerMixin):  # type: ignore
-    pass
+    def setup(self):
+        super().setup()
+        self.test_window = None
+        self.reuse_window = self.parent.reuse_window
+
+    def load_runner(self, url_protocol):
+        prev_runner_handle = self.runner_handle
+        super().load_runner(url_protocol)
+        if self.runner_handle != prev_runner_handle:
+            # Ensure the orphaned test window is closed.
+            self.close_test_window()
+
+    def close_test_window(self):
+        if self.test_window:
+            self._close_window(self.test_window)
+            self.test_window = None
+
+    def close_old_windows(self):
+        self.webdriver.actions.release()
+        for handle in self.webdriver.handles:
+            if handle not in {self.runner_handle, self.test_window}:
+                self._close_window(handle)
+        if not self.reuse_window:
+            self.close_test_window()
+        self.webdriver.window_handle = self.runner_handle
+        return self.runner_handle
+
+    def _close_window(self, window_handle):
+        try:
+            self.webdriver.window_handle = window_handle
+            self.webdriver.window.close()
+        except error.NoSuchWindowException:
+            pass
+
+    def open_test_window(self, window_id):
+        if self.test_window:
+            # Try to reuse the existing test window by emulating the `about:blank`
+            # page with no history you would get with a new window.
+            try:
+                self.webdriver.window_handle = self.test_window
+                # Reset navigation history with Chrome DevTools Protocol:
+                # https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-resetNavigationHistory
+                body = {
+                    "cmd": "Page.resetNavigationHistory",
+                    "params": {},
+                }
+                self.webdriver.send_session_command("POST", "goog/cdp/execute",
+                                                    body=body)
+                self.webdriver.url = "about:blank"
+                return
+            except error.NoSuchWindowException:
+                self.test_window = None
+        super().open_test_window(window_id)
+
+    def get_test_window(self, window_id, parent, timeout=5):
+        if self.test_window:
+            return self.test_window
+        # Poll the handles endpoint for the test window like the base WebDriver
+        # protocol part, but don't bother checking for the serialized
+        # WindowProxy (not supported by Chrome currently).
+        deadline = time.time() + timeout
+        while time.time() < deadline:
+            handles = self.webdriver.handles
+            if len(handles) == 2:
+                self.test_window = next(iter(set(handles) - {parent}))
```

Comment:
Should we assigned to member variable if this is a get function? Is this done purposefully?

Reply:
Yes, this state is necessary for `load_runner` and `close_old_windows` to distinguish the original test window from other windows that the test itself may have created. However, making the test window handle part of the interface would require overriding `WebDriverTestharnessExecutor.do_testharness`, which I think is worse than treating test window persistence as an implementation detail of the Chrome protocol implementation that's hidden from the executor. Added some comments to this effect.

(Even though it's not required, I should also note that `get_test_window` and `close_old_windows` are idempotent and work like a cache.)

Reply:
OK, I think I see what you are doing here. There are quite a bit duplicated code. Can we make a utility function, then  do `self.test_window = super()...` here?

Reply:
Done. The loops in `get_test_window` cannot be deduped because we want to skip over the somewhat expensive (~0.1s) `return window[...]` script call in the base executor that doesn't do anything in Chrome.